kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,603 @@
+#[derive(Clone, Debug, PartialEq)]
+enum State {
+    /// This client instance is the leader.
+    Leading {
+        /// The last observed lease state.
+        lease: Lease,
+        /// The last time this lease state updated as leader.
+        last_updated: Instant,
+    },
+    /// A state indicating that a different client is currently the leader, identified by the
+    /// encapsulated string.
+    ///
+    /// When a new leader is detected, this value will be updated with the leader's identity.
+    Following {
+        /// The ID of the current leader.
+        leader: String,
+        /// The last observed lease state.
+        lease: Lease,
+        /// The last time this lease state was observed.
+        last_updated: Instant,
+    },
+    /// A state indicating that the lease state is unknown, does not exist, or that the
+    /// corresponding leader elector task is starting or stopping.
+    Standby,
+}
+
+impl State {
+    /// Get a reference to the last known lease state.
+    fn get_lease(&self) -> Option<&Lease> {
+        match self {
+            Self::Leading { lease, .. } | Self::Following { lease, .. } => Some(lease),
+            Self::Standby => None,
+        }
+    }
+}
+
+/// Different states which a leader elector may be in.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum LeaderState {
+    /// This client instance is the leader.
+    Leading,
+    /// A state indicating that a different client is currently the leader.
+    Following,
+    /// A state indicating that the lease state is unknown, does not exist, or that the
+    /// corresponding leader elector task is starting or stopping.
+    Standby,
+}
+
+impl LeaderState {
+    /// Check if currently in `Leader` state.
+    #[allow(clippy::must_use_candidate)]
+    pub fn is_leader(&self) -> bool {
+        matches!(self, Self::Leading)
+    }
+}
+
+/// A handle to a leader elector task.
+pub struct LeaderElectorHandle {
+    /// Shutdown channel.
+    shutdown: oneshot::Sender<()>,
+    /// A watch signal over the observed leader state.
+    state: watch::Receiver<LeaderState>,
+    /// A join handle to the spawned leader elector task.
+    handle: JoinHandle<()>,
+}
+
+impl LeaderElectorHandle {
+    /// Get a handle to the state signal of this leader elector task.
+    ///
+    /// This signal receiver may be embedded in other parts of a program and used to govern actions
+    /// taken by the app in accordance with leader election state.
+    ///
+    /// Note that this leader elector task itself, as well as any other controller group members using
+    /// this same coordination implementation, will have an open watch on the lease object in the
+    /// Kubernetes API, and thus will see lease updates as they take place. This helps to ensure a
+    /// high degree of fencing, though it is not guaranteed.
+    #[allow(clippy::must_use_candidate)]
+    pub fn state(&self) -> watch::Receiver<LeaderState> {
+        self.state.clone()
+    }
```

Comment:
I'm wondering if there are some more ergonomic ways to expose this. Watching and matching doesn't feel like the most user focused setup atm. Some ideas;

- `async LeaderElectorHandle::ensure_leader(&self)`
- `async LeaderElectorHandle::status(&self) -> Arc<LeaderState>`

Reply:
Well, I've built multiple controllers / operators off of this code, and typically the useful properties which I have observed:
- We want to be able to react to leadership changes when they take place. This is provided via `.changed()` currently. We can then check the current value, and the application take take appropriate action based on its new role:
  - For leaders, this will typically mean start a new "full reconciliation".
  - For followers, this might mean a lot of different things. Maybe cancel some spawned tasks. Maybe purge some caches. This is very use-case dependent.
- Next, we want to actually know what state we are in. This is provided via `.borrow()` currently. As pointed out above, controllers will need to have some branching logic based on their current leadership state.

All in all, the current use of `watch::Receiver<_>` provides a single unified interface for the pertinent data. It allows for a single line in the controllers `select!` loop, and **VERY importantly,** it also allows for folks to check leadership state before performing reconciliation: `if !leader_state.borrow().is_leader() { return }` or the like.

That last bit is obviously an optimization, but it can help with fencing a bit. For example, here is a slightly modified snippet of what this looks like in one of the controllers/operators I maintain right now:

```rust
// Core event loop.
loop {
    tokio::select! {
        // .. lots of other elements here .. //
        _ = self.leader_state.changed() => {
            if self.leader_state.borrow_and_update().is_leader() {
                self.emit_reconcile_event(Reconcile::Full, None); // None is an optional delay.
            }
        }
    }
}
```

So, our use case is a bit too complex to use the controller/reconciler types in this crate directly, but this can be easily adapted (without code changes) to fit into those patterns.

Reply:
Resolved?